libraries/SITL/SIM_Sailboat.cpp

```diff
@@ -0,0 +1,180 @@
+    Aircraft(home_str, frame_str),
+    max_wheel_turn(35),
+    turning_circle(1.8)
+{
+}
+
+// calculate the lift and drag as values from 0 to 1
+// given an apparent wind speed in m/s and angle-of-attack in degrees
+void Sailboat::calc_lift_and_drag(float wind_speed, float angle_of_attack_deg, float& lift, float& drag)
+{
+    // check extremes
+    if (angle_of_attack_deg <= 0.0f) {
+        lift = lift_curve[0];
+        drag = drag_curve[0];
+        return;
+    }
+    if (angle_of_attack_deg >= 170.0f) {
+        lift = lift_curve[17];
+        drag = drag_curve[17];
+        return;
+    }
+
+    uint8_t index = constrain_int16(angle_of_attack_deg / 10, 0, 17);
+    float remainder = angle_of_attack_deg - (index * 10.0f);
+    lift = linear_interpolate(lift_curve[index], lift_curve[index+1], remainder, 0.0f, 10.0f);
+    drag = linear_interpolate(drag_curve[index], drag_curve[index+1], remainder, 0.0f, 10.0f);
+
+    // apply scaling by wind speed
+    lift *= wind_speed;
+    drag *= wind_speed;
+}
+
+/*
+  return turning circle (diameter) in meters for steering angle proportion in degrees
+*/
+float Sailboat::turn_circle(float steering)
+{
+    if (fabsf(steering) < 1.0e-6) {
+        return 0;
+    }
+    return turning_circle * sinf(radians(max_wheel_turn)) / sinf(radians(steering*max_wheel_turn));
+}
+
+/*
+   return yaw rate in degrees/second given steering_angle and speed
+*/
+float Sailboat::calc_yaw_rate(float steering, float speed)
+{
+    if (fabsf(steering) < 1.0e-6 or fabsf(speed) < 1.0e-6) {
+        return 0;
+    }
+    float d = turn_circle(steering);
+    float c = M_PI * d;
+    float t = c / speed;
+    float rate = 360.0f / t;
+    return rate;
+}
+
+/*
+  return lateral acceleration in m/s/s
+*/
+float Sailboat::calc_lat_accel(float steering_angle, float speed)
+{
+    float yaw_rate = calc_yaw_rate(steering_angle, speed);
+    float accel = radians(yaw_rate) * speed;
+    return accel;
+}
+
+/*
+  update the sailboat simulation by one time step
+ */
+void Sailboat::update(const struct sitl_input &input)
+{
+    // update wind
+    update_wind(input);
+
+    // in sailboats the steering controls the rudder, the throttle controls the main sail position
+    float steering = 2*((input.servos[0]-1000)/1000.0f - 0.5f);
+
+    // calculate mainsail angle from servo output 4, 0 to 90 degrees
+    float mainsail_angle_bf = constrain_float((input.servos[3]-1000)/1000.0f * 90.0f, 0.0f, 90.0f);
+
+    // calculate apparent wind in earth-frame (this is the direction the wind is coming from)
+    Vector3f wind_apparent_ef = wind_ef + velocity_ef;
+    const float wind_apparent_dir_ef = degrees(atan2f(wind_apparent_ef.y, wind_apparent_ef.x));
+    const float wind_apparent_speed = safe_sqrt(sq(wind_apparent_ef.x)+sq(wind_apparent_ef.y));
+
+    // calculate angle-of-attack from wind to mainsail
+    float aoa_deg = MAX(fabsf(wrap_180(wind_apparent_dir_ef - degrees(AP::ahrs().yaw))) - mainsail_angle_bf, 0);
```

Comment:
This shouldn't be looking into the AHRS. It gives a circular dependency on physics testing. Using SITL state is better.

Reply:
Indeed, I actually tried using sitl's "heading" variable but it's not correct because it's derived from the vehicle's velocity meaning it is only valid while the vehicle is moving.  So I'll leave this as-is until it becomes clear what SITL variable does hold the vehicle's actual yaw.